src/main/java/io/anserini/collection/AclCollection.java

```diff
@@ -0,0 +1,201 @@
+
+    ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+    try {
+      this.volumes = mapper.readValue(new File(path.toString(), "/volumes.yaml"), JsonNode.class);
+    } catch (IOException e) {
+      LOG.error("Unable to open volumes.yaml");
+      return;
+    }
+  }
+
+  @Override
+  public FileSegment<AclCollection.Document> createFileSegment(Path p) throws IOException {
+    return new Segment(p);
+  }
+
+  /**
+   * A file in a YAML collection for ACL papers containing multiple entires.
+   */
+  public class Segment extends FileSegment<AclCollection.Document> {
+    private Map.Entry<String, JsonNode> nodeEntry = null;
+    private Iterator<Map.Entry<String, JsonNode>> iter = null; // iterator for JSON document object
+
+    protected Segment(Path path) throws IOException {
+      super(path);
+
+      // read YAML file into JsonNode format
+      bufferedReader = new BufferedReader(new FileReader(path.toString()));
+      ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+      MappingIterator<JsonNode> iterator =
+        mapper.readerFor(JsonNode.class).readValues(bufferedReader);
+
+      if (iterator.hasNext()) {
+        JsonNode root = iterator.next();
+        iter = root.fields();
+        if (iter.hasNext()) {
+          nodeEntry = iter.next();
+        }
+      }
+    }
+
+    @Override
+    public void readNext() throws NoSuchElementException {
+      if (nodeEntry == null) {
+        throw new NoSuchElementException("JsonNode is empty");
+      } else {
+        bufferedRecord = new AclCollection.Document(nodeEntry);
+        if (iter.hasNext()) {
+          nodeEntry = iter.next();
+        } else {
+          atEOF = true; // there is no more JSON object in the bufferedReader
+        }
+      }
+    }
+  }
+
+  /**
+   * A document in a JSON collection.
+   */
+  public class Document implements SourceDocument {
+    private String id;
+    private String contents;
+    private JsonNode paper;
+    private JsonNode volume;
+    private List<String> authors;
+    private List<String> venues;
+    private List<String> sigs;
+
+    public Document(Map.Entry<String, JsonNode> jsonEntry) {
+      id = jsonEntry.getKey();
+      paper = jsonEntry.getValue();
+      contents = getOrDefault(paper, "title") + " " + getOrDefault(paper, "abstract_html");
+
+      // Process author facets
+      authors = new ArrayList<>();
+
+      if (paper.has("author")) {
+        ArrayNode authorNode = (ArrayNode) paper.get("author");
+        authorNode.elements().forEachRemaining(node -> {
+          authors.add(((ObjectNode) node).get("full").asText());
+        });
+      }
+
+      // Retrieve parent volume metadata
+      String parentVolumeId = paper.get("parent_volume_id").asText();
+      volume = AclCollection.this.volumes.get(parentVolumeId);
+
+      // Process venue facets
+      venues = new ArrayList<>();
+      ArrayNode venuesNode = (ArrayNode) volume.get("venues");
```

Comment:
If you're gonna do functional style, might as well go all the way and use `.collect(Collectors.toList());` or something similar? Same below.

Reply:
I think I'd need to convert the JsonNode iterator to a `Stream` first, which might be harder to read:

Reply:
I removed the curly braces and put everyone on one line, let me know if that's fine

Reply:
Yea, original looks better. Wish Java had better stream/iterator interop.